### ISSUE TITLE
Manage Terms and Conditions error

### ISSUE THREAD

**Issue by Marx15** [OpenMage/magento-lts#1975]
Hello everyone

I have a question.

I was currently still using OpenMage LTS 19.4.13.
Now I have noticed (you don't edit these often) that when I click on Sales> Terms and Conditions and maybe edit a new or an existing order condition, I get an error message:

There has been an error processing your request
Element with id "position" already exists
Trace: .............

do you have an idea?

Thanks very much.

**Comment by ADDISON74** [comment]
You should check this change merged into OpenMage code last year: https://github.com/OpenMage/magento-lts/pull/1480.

Analyze the **checkout_agreement table** in the database and the modified source code in PR. There may not have been a complete source code update. I checked with version 20.0.13 and I did not identify any issues in editing the terms and conditions.

In the future please be more explicit in reporting the problem. Mentioning if it is an installed version test from scratch or a modified one, if you have updated and how, what is the error shown in the logs on pages, what did you undertake in identifying and solving it. Thank you for your understanding.

**Comment by Marx15** [comment]
Thanks in advance.

This is an upgrade from magento 1 (last version) to OpenMage LTS 19.4.13.

The problem only arises in the backend when I want to create a new one or edit it.

I can't see anything in the database. How should it possibly look like that it is correct?

thank you :-)

**Comment by ADDISON74** [comment]
I hope you made such an upgrade in a test environment and you did not launch the website in production.

You can't say you don't see anything in the database. Did you find the **checkout_agreement** table? Did you look at its structure (colums)? That PR if you analyze it makes a change in this table and you must first determine if it was done correctly, then if the files are changed or not.

If you are in the test environment you can replace the following files with the initial ones from Magento and see what happens

/app/code/core/Mage/Adminhtml/Block/Checkout/Agreement/Edit/Form.php

/app/code/core/Mage/Adminhtml/Block/Checkout/Agreement/Grid.php

/app/code/core/Mage/Checkout/Block/Agreements.php

/app/code/core/Mage/Checkout/sql/checkout_setup/mysql4-upgrade-1.6.0.0-1.6.0.1.php

/app/code/core/Mage/Checkout/etc/config.xml

If you see the grid and clicking on a row and the error appears then it is the first file.

**Comment by ADDISON74** [comment]
You need detailed support to resolve the situation. Without access to the database and the file system it is difficult to identify what is happening through an exchange of messages.

**Comment by Marx15** [comment]
I could give you that ;-)

is there a possibility to have this table re-created? maybe this would solve the problem.

I'm not a professional. Sorry ;-)

**Comment by ADDISON74** [comment]
I am sorry I cannot help you with that maybe other could help you. 

The only thing I can help is making a comparison between your table structure and OpenMage, but you should export it (e.g. phpMyAdmin, mysqldump command) in SQL format and post it here. For the files I mentioned above you should use diff command and check the differences between initial Magento files and OpenMage. There are only a few files. Making a backup of them an replace should be another fast approach.

**Comment by Marx15** [comment]
Too bad. but thank you anyway. ;-)
Maybe someone else will give you a tip :-)

The comparison you mean would be from this table (s):
`checkout_agreement`
and maybe also:
`checkout_agreement_store`

with the original OpenMage table?

**Comment by ADDISON74** [comment]
As you are not a professional are you able to investigate the database into an interface like phpMyAdmin?

This is the structure of table named checkout_agreement in phpMyAdmin without that PR applied (Magento 1.9.4.5). 

![checkout_agreement](https://user-images.githubusercontent.com/8360474/149181178-d8eddb4d-8b0c-41a2-8b5f-e2412daac5c5.png)

As you may see there is no column named "position" which was introduced by the PR.  checkout_agreement_store is another table and that has nothing to do with your issue.

If the tables structure is OK the you can investigate the files I mentioned before. I have given you all the ingredients to have a starting point in investigating the issue.

**Comment by Marx15** [comment]
I have several positions:
![grafik](https://user-images.githubusercontent.com/22854944/149185080-b81f6adf-3fd1-4dfa-b7ab-e1816b40f850.png)

![grafik](https://user-images.githubusercontent.com/22854944/149185328-5b377e85-f685-4007-bdef-9c09ef90695f.png)

**Comment by ADDISON74** [comment]
If there is nothing confidential in your data can you click on Browse button an post a new screenshot here?

**Comment by sreichel** [comment]
> This is an upgrade from magento 1 (last version) to OpenMage LTS 19.4.13.

Do you use Firegento_Magesetup?

The "position" column was added there before it was introduced to Openmage.

(i'll update the update script to check for existing column)

**Comment by Marx15** [comment]
![Screen](https://user-images.githubusercontent.com/22854944/149206665-3f2ea04a-29b1-4039-bac3-79ef44f5445c.jpg)

**Comment by Marx15** [comment]
Yes. I use Firegento_Magesetup

**Comment by ADDISON74** [comment]
That image is for Structure section in phpMyAdmin not for Browse as I requested.

Just disable Firegento_Magesetup module and see if you can edit your Terms.

**Comment by Marx15** [comment]
in the test shop, if I deactivate Firegento_Magesetup then it works normally again.
so it has to have something to do with it, but what?

**Comment by ADDISON74** [comment]
4 columns were created in **checkout_agreement** table by Firegento_Magesetup position, agreement_type, is_required and revocation_product_type. Now things are clear where the error came from. 

If Sven hadn't mentioned that module we wouldn't have realized too quickly what the cause was. Mainly because the initial post does not provide information about the extensions/modules used, whether it is a test version or in production and much more. I doubt that an experienced person could solve the cause of a problem based on information such as _"I upgraded from Magento to OpenMage and I have this error ..."_.

**Comment by Marx15** [comment]
![Screen](https://user-images.githubusercontent.com/22854944/149208731-1c72623d-fdd6-48a0-a34f-04710e5d9815.jpg)

**Comment by ADDISON74** [comment]
Data are there unaltered which is good. Sven will tell you more about how to solve the issue because he knows Firegento and he also made the change in OpenMage. At first glance I see no issue related to the **position** field in the table. There is a conflict when you want the edit data in this table that comes from Firegento.

This is OpenMage
```
$installer = $this;
$installer->startSetup();

$connection = $installer->getConnection();
$table = $installer->getTable('checkout/agreement');
$column = 'position';

if (!$connection->tableColumnExists($table, $column)) {
    $connection->addColumn(
        $table,
        $column,
        array(
            'type'      => Varien_Db_Ddl_Table::TYPE_SMALLINT,
            'length'    => 2,
            'nullable'  => false,
            'default'   => 0,
            'comment'   => 'Agreement Position'
        )
    );
}

$installer->endSetup();
```

This is Firegento_MageSetup

```
$installer = $this;
$installer->startSetup();

if (version_compare(Mage::getVersion(), '1.6', '<')) {
    $installer->run("
        ALTER TABLE `{$installer->getTable('checkout/agreement')}`
        ADD `position` SMALLINT( 2 ) NOT NULL COMMENT 'Agreement Position'
    ");

} else {
    $installer->getConnection()->addColumn(
        $installer->getTable('checkout/agreement'),
        'position',
        array(
            'type'      => Varien_Db_Ddl_Table::TYPE_SMALLINT,
            'length'    => 2,
            'nullable'  => false,
            'default'   => 0,
            'comment'   => 'Agreement Position'
        )
    );
}

$installer->endSetup();
```

**Comment by ADDISON74** [comment]
This is the table structure in OpenMage 20.0.13 for checkout_agreement

![checkout_agreement](https://user-images.githubusercontent.com/8360474/149223409-ba04872c-a2b9-4972-adf9-424600d75267.jpg)

An here is the SQL dump for table structure

```sql
SET SQL_MODE = "NO_AUTO_VALUE_ON_ZERO";
SET AUTOCOMMIT = 0;
START TRANSACTION;
SET time_zone = "+00:00";


/*!40101 SET @OLD_CHARACTER_SET_CLIENT=@@CHARACTER_SET_CLIENT */;
/*!40101 SET @OLD_CHARACTER_SET_RESULTS=@@CHARACTER_SET_RESULTS */;
/*!40101 SET @OLD_COLLATION_CONNECTION=@@COLLATION_CONNECTION */;
/*!40101 SET NAMES utf8mb4 */;

--
-- Database: `openmage_2013`
--

-- --------------------------------------------------------

--
-- Table structure for table `checkout_agreement`
--

CREATE TABLE `checkout_agreement` (
  `agreement_id` int(10) UNSIGNED NOT NULL COMMENT 'Agreement Id',
  `name` varchar(255) DEFAULT NULL COMMENT 'Name',
  `content` text COMMENT 'Content',
  `content_height` varchar(25) DEFAULT NULL COMMENT 'Content Height',
  `checkbox_text` text COMMENT 'Checkbox Text',
  `is_active` smallint(6) NOT NULL DEFAULT '0' COMMENT 'Is Active',
  `is_html` smallint(6) NOT NULL DEFAULT '0' COMMENT 'Is Html',
  `position` smallint(6) NOT NULL DEFAULT '0' COMMENT 'Agreement Position'
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COMMENT='Checkout Agreement';

--
-- Indexes for dumped tables
--

--
-- Indexes for table `checkout_agreement`
--
ALTER TABLE `checkout_agreement`
  ADD PRIMARY KEY (`agreement_id`);

--
-- AUTO_INCREMENT for dumped tables
--

--
-- AUTO_INCREMENT for table `checkout_agreement`
--
ALTER TABLE `checkout_agreement`
  MODIFY `agreement_id` int(10) UNSIGNED NOT NULL AUTO_INCREMENT COMMENT 'Agreement Id';
COMMIT;

/*!40101 SET CHARACTER_SET_CLIENT=@OLD_CHARACTER_SET_CLIENT */;
/*!40101 SET CHARACTER_SET_RESULTS=@OLD_CHARACTER_SET_RESULTS */;
/*!40101 SET COLLATION_CONNECTION=@OLD_COLLATION_CONNECTION */;
```

**Comment by sreichel** [comment]
Same issue popped up here https://github.com/firegento/firegento-magesetup/pull/355

I've added "position" field to both ... first Magesetup, and later to Openmage - w/o checking for already existing field.

I'll (try to) fix it tomorrow.

**Comment by sreichel** [comment]
Please update Magesetup .... https://github.com/firegento/firegento-magesetup/pull/355

**Comment by Marx15** [comment]
Does it mean wait for an update from Magesetup to 2.7.0? and if something has to be changed in the "Terms and Conditions" item, then simply deactivate the Firegento_Magesetup module, edit it and activate it again?

**Comment by ADDISON74** [comment]
According to this page, Firegento_Magesetup team released 10 days ago a new version 2.7.0.

https://github.com/firegento/firegento-magesetup/releases/tag/2.7.0

Check your current version and if it is bellow upgrade to 2.7.0 as Sven pointed before.

**Comment by Marx15** [comment]
Thank you very much for your hint.
I had seen it too, but it is exactly in the original (new) file, under:
  firegento-magesetup/src/app/code/community/FireGento/MageSetup/etc/config.xml

still:
     <modules>
         <FireGento_MageSetup>
             <version>2.6.0</version>
         </FireGento_MageSetup>
     </modules>

got me a bit irritated. sorry :-)

**Comment by ADDISON74** [comment]
Although they mention that they have released a new version 2.7.0, from the configuration file it is still version 2.6.0. but the change Sven showed you above must be made. Whether it is 2.7 or 2.6 does not affect anything as long as no SQL script is executed. Make that change in that file and see what is happening. Also, you can download the entire archive and copy the files over the existing ones.

Here is the modification you have to do in your file using an editor

https://github.com/firegento/firegento-magesetup/pull/355/files

I'm sorry to say, but you're already requesting technical support at the lowest level and it has nothing to do with the error in the OpenMage code.

**Comment by Marx15** [comment]
I copied and pasted the file.
it worked. Many Thanks.

I just wanted to draw your attention to this if something has been forgotten. sorry ;-)
